# Linear Programming Practice (Simplex Method)

## Minimize the costs for blending orange juice

First let's read the specs provided from the different juice suppliers, I am using pandas only to visualize the table automatically

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('Supplier Specs.csv')

df

Varietal      Region  Qty Available (1,000 Gallons)  \
0           Hamlin      Brazil                            672   
1          Mosambi       India                            400   
2         Valencia     Florida                           1200   
3           Hamlin  California                            168   
4          Gardner     Arizona                             84   
5          Sunstar       Texas                            210   
6         Jincheng       China                            588   
7            Berna       Spain                            168   
8            Verna      Mexico                            300   
9   Biondo Commune       Egypt                            210   
10      Belladonna       Italy                            180   

    Brix / Acid Ratio  Acid (%)  Astringency (1-10 Scale)  Color (1-10 Scale)  \
0                10.5    0.0060                         3                   3   
1                 6.5    0.0140                         7                   1   
2                12.0    0.0095                         3                   3   
3                11.0    0.0100                         3                   5   
4                12.0    0.0070                         1                   5   
5                10.0    0.0070                         1                   5   
6                 9.0    0.0135                         7                   3   
7                15.0    0.0110                         4                   8   
8                 8.0    0.0130                         8                   3   
9                13.0    0.0130                         3                   5   
10               14.0    0.0050                         3                   9   

    Price (per 1K Gallons)  Shipping (per 1K Gallons)  
0                      500                        100  
1                      310                        150  
2                      750                          0  
3                      600                         60  
4                      600                         75  
5                      625                         50  
6                      440                        120  
7                      600                        110  
8                      300                         90  
9                      460                        130  
10                     505                        115

* Brix/Acid ratio: Brix is a measure of sweetness in the juice, so Brix/Acid ratio is a
measure of sweetness to tartness, which in the end, is really what orange juice is
all about.
* Acid (%): Acid as a percentage of the juice is broken out individually, because at
a certain point, it doesn’t really matter how sweet the juice is, it’s still too acidic.
* Astringency (1–10 scale): A measure of the “green” quality of the juice. It’s that
bitter, unripe, planty fl avor that can creep in. This scale is assessed by a panel of
tasters at each juicing facility on a scale of 1–10.

## Our requirements for the optimization are the following:

* Lowest cost purchase plan for January, February, and March that meets a projected demand of 600,000 gallons of juice in January and February and 700,000 gallons in March.
* Our company has entered an agreement with the state of Florida which provides the company tax incentives so long as the company buys at least 40 percent of its juice each month from Florida Valencia growers. Under no circumstances are you to violate this agreement.
* The Brix/Acid ratio (BAR) must stay between 11.5 and 12.5 in each month’s blend.
* The acid level must remain between 0.75 and 1 percent.
* The astringency level must stay at 4 or lower.
* Color must remain between 4.5 and 5.5. Not too watery, not too dark.

## Real quickly shove those requirements into an outline of an LP formulation:
* **Objective**: Minimize procurement costs.
* **Decisions**: Amount of each juice to buy each month
* **Constraints**:
    * Demand
    * Supply
    * Florida Valencia requirement
    * Flavor
    * Color

# Formulating the problem

I will denote the variables which we need to an optimal solution for as $x_0, x_1, ...$. For this problem these are the quantities of juice which we need to buy from each supplier. In our case we have 11 different suppliers which results in 11 variables, named from $x_0$ to $x_{10}$. Let's assume that $x_0$ is the juice from Brazil, $x_1$ is the juice from India and etc.



## Objective function

The aim which we have is to minimize the total cost (purchase + shipping). Well the cost depends solely on the amount of juice which we buy, thus we can write that the total cost for buying 1000 galons of juice from Brazil will be $500 + 100 = 600$. I didn't write 1000 anywhere as the purchase and shipping costs are priveded per 1000 galons.

Having all this in mind we can write the objective function in the following form:

$(500+100)x_0 + (310+150)x_1 + 750x_2 + (600+60)x_3 + (600+75)x_4 + (625+50)x_5 + (440+120)x_6 + (600+110)x_7 + (300+90)x_8 + (460+130)x_9 + (505+115)x_{10}$

And for the sake of clarity once again:

$f = 600x_0 + 460x_1 + 750x_2 + 660x_3 + 675x_4 + 675x_5 + 560x_6 + 710x_7 + 390x_8 + 590x_9 + 620x_{10}$

$$f = \sum{((purchase_i+shipping_i)x_i)}$$

## Demand (equality constraint)

In the problem we also have to account for total amount of juice which we blend:

* Lowest cost purchase plan for January, February, and March that meets a projected demand of 600,000 gallons of juice in January and February and 700,000 gallons in March.

This translates into the following equalities:

$ D = \sum{x_i} = 600 $ - for January and February

$ D = \sum{x_i} = 700 $ - for March

## Supply (bounds)

On the other hand in the supplier table we see that we do not have a limited supply of $x_0, x_1, x_2, ...$ This fact imposes the following requirements on our variables:

$$x_0 \le 672$$
$$x_1 \le 400$$
$$x_2 \le 1200$$
$$x_3 \le 168$$
$$x_4 \le 84$$
$$x_5 \le 210$$
$$x_6 \le 588$$
$$x_7 \le 168$$
$$x_8 \le 300$$
$$x_9 \le 210$$
$$x_{10} \le 180$$

Authough it looks good these limits are not sufficient for simulating a realistic scenario. Imagine that we decide to purchase negative amout of juice form some/all suppliers. In addition in the problem statement we see that each month we need to purchase certain amount of juice depending on our production (40%) so we can keep the negotiated tax benefits. Combining these two points we result with the following limits for our variables:

$$0 \le x_0 \le 672$$
$$0 \le x_1 \le 400$$
$$0.4D \le x_2 \le 1200$$
$$0 \le x_3 \le 168$$
$$0 \le x_4 \le 84$$
$$0 \le x_5 \le 210$$
$$0 \le x_6 \le 588$$
$$0 \le x_7 \le 168$$
$$0 \le x_8 \le 300$$
$$0 \le x_9 \le 210$$
$$0 \le x_{10} \le 180$$

## Flavor and Color (inequality constraints)

The last point which we need to take into account is the quality of the juice which we will blend. Let's start with the first criteria:

* **The Brix/Acid ratio (BAR) must stay between 11.5 and 12.5 in each month’s blend.**

This can be formulated as the weighted avarage between the amount of juice and their BAR to be within the rage $[11.5;12.5]$ or:

$$11.5 \le \frac{\sum{(BAR_i*x_i)}}{\sum{x_i}} \le 12.5$$

The next point is:

* **The acid level must remain between 0.75 and 1 percent.**

Following the logic from the previous point, this can be written as:

$$0.0075 \le \frac{\sum{(Acid_i*x_i)}}{\sum{x_i}} \le 0.01$$

The next point is:

* **The astringency level must stay at 4 or lower**

which can be written as:

$$0 \le \frac{\sum{(Astringency_i*x_i)}}{\sum{x_i}} \le 4.0$$

and the last point is:

* **Color must remain between 4.5 and 5.5. Not too watery, not too dark.**

which can be written as:

$$4.5 \le \frac{\sum{(Color_i*x_i)}}{\sum{x_i}} \le 5.5$$

These inequalities are easy to read but we would have some trouble implementing them in Python directly as we only need the coefficients in front of each $x_i$. Thus we need to rewrite them in the following general form:

$$min \le \frac{\sum{(k_ix_i)}}{\sum{x_i}} \le max$$

$$min\sum{x_i} \le \sum{(k_ix_i)} \le max\sum{x_i}$$

In other words if we want to keep our coefficients as they are given in the supplier table we just need to scale our limits for the equalities with the current total amount of juice which we blend per month.

# Translating this into Python

I am going to use SciPy build in optimization method called: linprog

Before I actually start with the implementation you can check its documentation over here: [linprog](https://docs.scipy.org/doc/scipy/reference/optimize.linprog-simplex.html). This method allows only upper bound inequality constrains. In our case we also have lower bound inequality constrains. One easy workaround is to present $a <= b$ as $-a >= -b$

Ok Let's start!

In [4]:
from scipy.optimize import linprog

def fromPandas(dataframe, colName):
    return dataframe[[colName]].values.reshape(1,11)[0]

## A_ub * x <= b_ub - inequality constraint
## A_eq * x == b_eq - equality constraint

A_eq = np.array([
    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
]) # coefficient matrix for the equality constraints

u_eq = [600.0] # demand

## reading the actual numbers from the pandas dataframe and then converting them to vectors
BAR = fromPandas(df, 'Brix / Acid Ratio')
acid = fromPandas(df, 'Acid (%)')
astringency = fromPandas(df, 'Astringency (1-10 Scale)')
color = fromPandas(df, 'Color (1-10 Scale)')
price = fromPandas(df, 'Price (per 1K Gallons)')
shipping = fromPandas(df, 'Shipping (per 1K Gallons)')
upperBounds = fromPandas(df, 'Qty Available (1,000 Gallons)')

lowerBounds = [0]*len(upperBounds) # list with length 11 and value 0
lowerBounds[2] = 0.4*u_eq[0] # adding the Florida tax bound

bnds = [(0,0)]*len(upperBounds) # bounds vector
for i in range(0,len(upperBounds)):
    bnds[i] = (lowerBounds[i], upperBounds[i])

c = price + shipping # objective function coefficients vector

A_ub = np.array([
        BAR, 
        acid, 
        astringency, 
        color, 
        -BAR, 
        -acid, 
        -astringency, 
        -color
]) # coefficient matrix for inequality constraints

b_ub = np.array([12.5, 0.01, 4.0, 5.5, -11.5, -0.0075, -0.0, -4.5]) # limits for the inequalities

b_ub = b_ub * u_eq[0] # scaling the inequalities limits with the demand

result = linprog(c, A_ub, b_ub, A_eq, u_eq, bounds=(bnds))

xOptimized = result.x

def objFunc(x,c):
    # returns the value of our objective function for a variable vector x and coefficient vector c
    return round(sum(x*c),0)

def weightedAverage(a,b):
    # returns the weighted avarage of the vector a and b with respect to a
    return sum(a*b)/sum(a)

print("Objective function value: ", objFunc(xOptimized, c), "\n")

from IPython.display import HTML, display
from tabulate import tabulate

# supplier segmentation table
print("\n Supplier Segmentation Display")
header = ['Region', 'Quantity ordered']
region = fromPandas(df, 'Region')
content = np.stack((region, xOptimized), axis=-1)

display(HTML(tabulate(content , headers=header, tablefmt='html')))

# juice quality table
header = ['Property', 'min', 'Our juice', 'max']
content = [
    ['Brix / Acid Ratio', 11.5   , weightedAverage(xOptimized,BAR)        , 12.5],
    ['Acid'             , 0.0075 , weightedAverage(xOptimized,acid)       , 0.01],
    ['Astringency'      , 0.0    , weightedAverage(xOptimized,astringency), 4.0],
    ['Color'            , 4.5    , weightedAverage(xOptimized,color)      , 5.5]
]

print("\n Juice Quality Display")
display(HTML(tabulate(content , headers=header, tablefmt='html')))

Objective function value:  371724.0 


 Supplier Segmentation Display



 Juice Quality Display


## Result

We got the exact result as in excel.

![title](img/Jan.png)